hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -272,56 +255,47 @@ private void collectNestedTypes(final QueryDataType dataType, final Map<String,
 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.converter = Converters.getConverter(in.readInt());
-        if (!converter.getTypeFamily().equals(QueryDataTypeFamily.OBJECT)) {
+        converter = Converters.getConverter(in.readInt());
+        if (converter.getTypeFamily() != QueryDataTypeFamily.OBJECT) {
             return;
         }
 
-        this.objectTypeKind = in.readInt();
-        this.objectTypeName = in.readString();
-        this.objectTypeMetadata = in.readString();
+        objectTypeKind = TypeKind.of(in.readInt());
```

Comment:
eager conversion to enum might be risky from the point of view of compatibility. This will fail earlier (on load, not on use).
in addition, `TypeKind.of` is broken, it uses `ordinal` instead of `value`

Reply:
`TypeKind.of` is currently not broken. If we remove a `TypeKind` in the future, the implementation should be changed to look for `value` instead of `ordinal`. What is the advantage of failing later? Is there any possibility of such a failure?

Reply:
`ordinal` depends on the order of constants in source code which should not be the base for binary formats. "broken" was a shortcut, I meant that is uses risky mechanism. Currently because `value` is the same as `ordinal` the code works.

Reply:
> What is the advantage of failing later? Is there any possibility of such a failure?

It is important for forward compatibility.
`TableResolverImpl.getTables` reads all tables eagerly (which is not the best approach but that is how it works). This in turn deserializes QueryDataTypes. If we ever add new `TypeKind`, during rolling minor version upgrade queries on old version might fail (they should be executed on subset of the cluster) if someone creates UDT with new type on the upgraded part of the cluster.

Reply:
there might be a similar problem if we decide to remove one of the TypeKinds, but this time in the newer members of the cluster.

Reply:
but the scenario with creating UDTs and MAPPINGs during RU does not seem very probable and we generally should discourage users from doing that, but AFAIK it is not forbidden. @ivanthescientist  WDYT?

Reply:
AFAIK we have no inter-member forward compatibility guarantees. It only applies to client-to-member and MC-to-member communication.